### PR TITLE
Convert `bool_and` & `bool_or` to UDAF

### LINKED ISSUE #11008: Convert `BoolAndOr` to UDAF
### Is your feature request related to a problem or challenge?

Part of #8708. 

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jcsherin**
## Which issue does this PR close?

Closes #11008.

## What changes are included in this PR?

* Converts `bool_and`, `bool_or` to UDAF
* Export above as fluent style API for creating `Expr`s
* Add expressions to `roundtrip_expr_api` test

## Are these changes tested?
* [Existing sqllogictest in `aggregate.slt`](https://github.com/apache/datafusion/blob/c6b2efccf6238cc87f2414efb28ae3b263ed27af/datafusion/sqllogictest/test_files/aggregate.slt#L3547-L3586)
* Logical plan round trip test

### CODE REVIEW

File: datafusion/functions-aggregate/src/bool_and_or.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+//         generic_test_op!(a, DataType::Boolean, BoolOr, ScalarValue::from(false))
+//     }
+//
+//     #[test]
+//     fn bool_or_all_nulls() -> Result<()> {
+//         let a: ArrayRef = Arc::new(BooleanArray::from(vec![None, None]));
+//         generic_test_op!(a, DataType::Boolean, BoolOr, ScalarValue::Boolean(None))
+//     }
+// }
+//
```
Comment (jcsherin) [inline datafusion/functions-aggregate/src/bool_and_or.rs:408]: This test code can be safely removed?

This seems to already [covered by existing sqllogictests here](https://github.com/apache/datafusion/blob/c6b2efccf6238cc87f2414efb28ae3b263ed27af/datafusion/sqllogictest/test_files/aggregate.slt#L3547-L3586).

File: datafusion/functions-aggregate/src/bool_and_or.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+use datafusion_common::internal_err;
+use datafusion_common::{downcast_value, not_impl_err};
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use datafusion_expr::function::{AccumulatorArgs, StateFieldsArgs};
+use datafusion_expr::utils::{format_state_name, AggregateOrderSensitivity};
+use datafusion_expr::{
+    Accumulator, AggregateUDFImpl, GroupsAccumulator, ReversedUDAF, Signature, Volatility,
+};
+
+use datafusion_physical_expr_common::aggregate::groups_accumulator::bool_op::BooleanGroupsAccumulator;
```
Comment (jcsherin) [inline datafusion/functions-aggregate/src/bool_and_or.rs:38]: I could use some help here on how to proceed with extracting `BooleanGroupsAccumulator` from `physical-expr-common`.

The `BooleanGroupsAccumulator` depends on `NullState` which in turn has other users as seen here:
```sh
$ rg NullState -c
datafusion-examples/examples/advanced_udaf.rs:4
datafusion/physical-expr/src/lib.rs:1
datafusion/physical-expr/src/aggregate/average.rs:3
datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs:2
datafusion/physical-expr-common/src/aggregate/groups_accumulator/bool_op.rs:4
datafusion/physical-expr-common/src/aggregate/groups_accumulator/prim_op.rs:4
datafusion/physical-expr-common/src/aggregate/groups_accumulator/accumulate.rs:15
```

File: datafusion/functions-aggregate/src/bool_and_or.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        Ok(DataType::Boolean)
+    }
+
+    fn accumulator(&self, _: AccumulatorArgs) -> Result<Box<dyn Accumulator>> {
+        Ok(Box::<BoolAndAccumulator>::default())
+    }
+
+    fn state_fields(&self, args: StateFieldsArgs) -> Result<Vec<Field>> {
+        Ok(vec![Field::new(
+            format_state_name(args.name, "bool_and"),
```
Comment (jcsherin) [inline datafusion/functions-aggregate/src/bool_and_or.rs:138]: Should the second argument here be `"acc"` which is the name of the field in `BoolAccumulator`?

File: datafusion/functions-aggregate/src/bool_and_or.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+    fn create_sliding_accumulator(
+        &self,
+        _: AccumulatorArgs,
+    ) -> Result<Box<dyn Accumulator>> {
+        Ok(Box::<BoolAndAccumulator>::default())
+    }
+
+    fn order_sensitivity(&self) -> AggregateOrderSensitivity {
+        AggregateOrderSensitivity::Insensitive
```
Comment (jcsherin) [inline datafusion/functions-aggregate/src/bool_and_or.rs:176]: The default is `AggregateOrderSensitivity::HardRequirement`. Is the use of `Insensitive` instead correct for `bool_and` and `bool_or`?

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 5]
```diff
@@ -671,6 +673,8 @@ async fn roundtrip_expr_api() -> Result<()> {
         bit_or(lit(2)),
         bit_xor(lit(2)),
         string_agg(col("a").cast_to(&DataType::Utf8, &schema)?, lit("|")),
+        bool_and(lit(true)),
+        bool_or(lit(true)),
```
Comment (jcsherin) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:677]: Added to `roundtrip_expr_api`.

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 2 above)
Comment (jcsherin) [reply datafusion/functions-aggregate/src/bool_and_or.rs:38]: It's also imported by the following in `functions-aggregate`:
1. `count`
2. `bit_and_or_xor`
3. `sum`

So maybe this is better tackled in a separate PR and is ok for now :thinking:

Comment (alamb) [reply datafusion/functions-aggregate/src/bool_and_or.rs:38]: I think we should leave `BooleanGroupsAccumulator` in `physical-expr-common` until we have moved the other boolean aggregate functionss over  -- then I think `BooleanGroupsAccumulator` should be able to move without issue

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 3 above)
Comment (alamb) [reply datafusion/functions-aggregate/src/bool_and_or.rs:138]: I think you can use `self.name()` here and be consistent with https://github.com/apache/datafusion/blob/0f80b9261f389241f1d76449940faccf7e237e80/datafusion/functions-aggregate/src/bit_and_or_xor.rs#L202

Hard codeing bool_and is ok too

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 4 above)
Comment (alamb) [reply datafusion/functions-aggregate/src/bool_and_or.rs:176]: `Insensitive` makes sense to me -- @mustafasrepo  perhaps you can confirm?

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 1 above)
Comment (alamb) [reply datafusion/functions-aggregate/src/bool_and_or.rs:400]: I agree these tests are redundant and can be removed (rather than commented out)

**Review by alamb — COMMENTED** [review]
Thank you @jcsherin  -- this is a really neat first PR

I think we should remove the commented out tests, but then this looks good to go from my perspective

cc @jayzhan211

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 1 above)
Comment (jcsherin) [reply datafusion/functions-aggregate/src/bool_and_or.rs:400]: Thanks for the confirmation. I'll remove the commented out tests.

File: datafusion/functions-aggregate/src/bool_and_or.rs (hunk 3 above)
Comment (jcsherin) [reply datafusion/functions-aggregate/src/bool_and_or.rs:138]: Will make this change.

**Comment by jcsherin** [comment]
I've pushed the following changes based on the code review:
- Delete the redundant test code
- Use `self.name()` instead of hard-coding either `bool_and`/`bool_or`

**Review by alamb — APPROVED** [review]
Thank you @jcsherin  -- looks great to me. I'll wait a while to merge this to let @jayzhan211  / @mustafasrepo  have a chance to review if they want

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by alamb** [comment]
Thanks again @jcsherin  and @jayzhan211

**Pull request merged**